### PR TITLE
feat: add intersection schema

### LINKED ISSUE #113: Missing `intersection` helper
I noticed there is a `union` helper but no `intersection` helper. Is this intended or just missing? If it's intended, how do you recommend creating an intersection? I tried to use `merge`, but it does not allow the same types to be merged that TypeScript allows.

### PR DESCRIPTION
**Pull request by alonidiom**
Fixes #113 (including the specific case mentioned)

### CODE REVIEW

**Comment by fabian-hiller** [comment]
Thank you for your contribution! The next few days I have to focus on my bachelor thesis. As soon as I find time, I'll get back to you.

**Comment by fabian-hiller** [comment]
If there is interest in this PR, please check [this comment](https://github.com/fabian-hiller/valibot/issues/113#issuecomment-1712312002) beforehand and give me feedback on it.

File: library/src/schemas/intersection/intersection.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+      return !issues && output
+        ? {
+            output: output.reduce((acc, value) => {
+              if (typeof value === 'object') {
+                return { ...acc, ...value };
+              }
+              return value;
+            }) as IntersectionOutput<TIntersectionOptions>,
+          }
+        : getIssues(
```
Comment (Karakatiza666) [inline library/src/schemas/intersection/intersection.ts:104]: In my case validation errors from intersected objects didn't bubble up, and I had to do this:
```
return !issues && output
        ? {
           ...
          }
        : issues?.length
            ? { issues }
            : getIssues(info, 'type', 'intersection', error || 'Invalid type', input, issues)
```
Please check this, and if valid this applies to intersectionAsync, too

File: library/src/schemas/intersection/intersection.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+      let output: [Output<IntersectionOptions[number]>] | undefined;
+
+      // Parse schema of each option
+      for (const schema of intersection) {
+        const result = schema._parse(input, info);
+
+        // If there are issues, set output and break loop
+        if (result.issues) {
+          issues = result.issues;
+          break;
```
Comment (Karakatiza666) [inline library/src/schemas/intersection/intersection.ts:84]: In my case validation errors from intersected objects didn't bubble up, and I had to do this:
```
        if (result.issues) {
          issues = result.issues
          while (issues?.length) {
            issues = (issues => issues.length ? issues as Issues : undefined)(issues.flatMap(i => i.issues ?? []))
              ?? issues
            break
          }
          break
        } else ...
```
Please check this, and if valid this applies to intersectionAsync, too

**Comment by fabian-hiller** [comment]
I will probably review this PR in the next few days.

**Comment by Karakatiza666** [comment]
@alonidiom Please check out my fork of your PR! https://github.com/Karakatiza666/valibot/commit/fd1830ff0d474c4b201e1aca05566c746b550371
I believe it improves type inference, and behaviour when handling parsing errors

**Comment by fabian-hiller** [comment]
I am currently working on this PR. I'll check your code in a moment.

**Comment by fabian-hiller** [comment]
`union` is a logical OR. For this reason, the issues are added as subissues, as they usually contradict each other. `intersection` is a logical AND, so in my opinion this procedure is not necessary here. I have also investigated the implementation of Zod in this regard.

I'm not done yet, but I welcome feedback or questions now before I finalize it.

**Comment by fabian-hiller** [comment]
Apart from that, I also revised the code that merges two or more outputs so that it should now work with every schema.

**Comment by Karakatiza666** [comment]
Will check if I encounter issues with new code tomorrow, wasn't able to try earlier

**Comment by fabian-hiller** [comment]
I have finalized the implementation. Feel free to test out the changes and let me know if anything can be improved.

**Comment by fabian-hiller** [comment]
I plan to merge this PR today or tomorrow.

**Pull request merged**